exporter: New metric for report ceph daemons health (#60167)
ceph/ceph: doc/monitoring/index.rst

It is good to outline that the main tool allowing users to observe and monitor a Ceph cluster is the **Ceph dashboard**. It provides graphics where the most important cluster and service metrics are represented. Most of the examples in this document are extracted from the dashboard graphics or extrapolated from the metrics exposed by the Ceph dashboard.

Ceph daemon health metrics
==========================

> The Ceph exporter provides a metric called ``ceph_daemon_socket_up`` that reports the health status of all Ceph daemons exposing an admin socket. These include the main Ceph daemons, encompassing any daemon capable of exposing performance counters.

Reviewer: ```suggestion
The Ceph exporter provides a metric called ``ceph_daemon_socket_up`` that reports the liveness status of each Ceph daemon that exposes an admin socket. 
```

> The Ceph exporter provides a metric called ``ceph_daemon_socket_up`` that reports the health status of all Ceph daemons exposing an admin socket. These include the main Ceph daemons, encompassing any daemon capable of exposing performance counters.

Reviewer 2: I understand the idea of using the socket status to see if the daemon is OK or not, but I'd recommend having a specific health field. A daemon can be reachable (network is OK) but not functioning because of XYZ (that would depend on each case). I'd recommend separating the health status of the daemon from its ability to communicate and have a separate metrics for both of them. For example in k8s world you have the **Liveness Probe** to see if the daemon is reachable or not and the **Readiness Probe** to see if the pod is ready do what it has to do.

Reviewer 2: > Maybe it was not a good idea to use the caption "ceph daemon health metrics". In fact I have named the new metric "ceph_daemon_socket_up" ... which i think clearly indicates what is used for.
> 
> The thing is that if you have this metric returned with a value of 0, that clearly indicates that the daemon is not working properly, and therefore is not healthy (that is why used the "title ceph daemon health metrics", but we can change it for "Ceph daemon readiness metrics". In any case, how do you suggest to use as name for the new metric, and what is your specific proposal about the "titles" or explanation in the documentation, should we talk about readiness/liveness?
> 
> This modification, comes from the need to have something to indicate that an RBD mirror daemon was not working. ) [rhbz#2146728](https://bugzilla.redhat.com/show_bug.cgi?id=2146728)): In fact, i think this metric, is the best way to know that something wrong has happened with the daemon, due to the fact that the Ceph exporter communicates directly with the daemon through the admin socket, and it is not affected by any network issue. With this metric you can detect very easily a problem in the daemon, when the issue happened and if still is happening. Without this metric, to know abut the daemon health/readiness/liveness you need to use a combination of queries using "Ceph health" plus different set of metrics depending on the daemon in order to obtain the same information ( when the daemon stop responding and during how long.)
> 
> There is border use case, which is that the daemon does not answer request through admin socket, but still it answers using other communication ways.. (http/https for example). But even in this case, to have the metric "ceph_daemon_socket_up" reporting the problem is useful, because that means that we are not going to have any "perf counter dump" from the daemon, leaving blindly the daemon monitoring.

I see, thank you @jmolmo for the detailed explanation. Now I understand better what this metric is and what's the purpose of its usage. I'd say it seems more related to readiness (as if the daemon admin socket is not working, we can't send any admin cmds) that liveness (as the daemon could still be working and processing requests). The name you choose for the metric `ceph_daemon_socket_up` I think is clear and describes what it's reporting (not sure if it would be more clear by adding _admin so we end up with `ceph_daemon_admin_socket_up` but that's a subtle detail).

> The ``ceph_daemon_socket_up`` metric indicates the health status of a Ceph daemon based on its ability to communicate through the admin socket, where a value of ``1`` means healthy, and ``0`` means unhealthy. Although a Ceph daemon might still be "alive" when it reports ``ceph_daemon_socket_up=0``, this situation highlights a significant issue in its functionality. As such, this metric serves as an excellent tool for detecting problems in any of the main Ceph daemons.

Reviewer: ```suggestion
The ``ceph_daemon_socket_up`` metric indicates the health status of a Ceph daemon based on its ability to respond via the admin socket, where a value of ``1`` means healthy, and ``0`` means unhealthy. Although a Ceph daemon might still be "alive" when it reports ``ceph_daemon_socket_up=0``, this situation highlights a significant issue in its functionality. As such, this metric serves as an excellent tool for detecting problems in any of the main Ceph daemons.
```

> The ``ceph_daemon_socket_up`` metric indicates the health status of a Ceph daemon based on its ability to communicate through the admin socket, where a value of ``1`` means healthy, and ``0`` means unhealthy. Although a Ceph daemon might still be "alive" when it reports ``ceph_daemon_socket_up=0``, this situation highlights a significant issue in its functionality. As such, this metric serves as an excellent tool for detecting problems in any of the main Ceph daemons.

Reviewer 3: Could `ceph_daemon_socket_up=1` (healthy) also result in unresponsive/unreachable daemons (e.g.: routing, network issues, firewalls, etc)? Meaning that we could have both false negatives and false positives?

Author: The metric only says that the daemon is not "answering" though the admin socket. 
That does not means that the daemon is "broken", is is not healthy but maybe it can continue doing his work. In any case is a situation that needs attention.

Labels:
- **``ceph_daemon``**: Identifier of the Ceph daemon exposing an admin socket on the host.
- **``hostname``**: Name of the host where the Ceph daemon is running.

Example:

.. code-block:: bash

   ceph_daemon_socket_up{ceph_daemon="mds.a",hostname="testhost"} 1
   ceph_daemon_socket_up{ceph_daemon="osd.1",hostname="testhost"} 0

> PromQL Expression to Detect Unresponsive Daemons in the Last 12 Hours:

Reviewer: ```suggestion
```

> To identify which Ceph daemons were not responsive at any point in the last 12 hours, you can use the following expression:

Reviewer: ```suggestion
To identify any Ceph daemons that were not responsive at any point in the last 12 hours, you can use the following PromQL expression:
```

.. code-block:: bash

>    ceph_daemon_socket_up == 0 or min_over_time(ceph_daemon_socket_up[12h]) == 0

Reviewer 3: Aren't we adding any alerts/Grafana Dashboard widgets for this?

Author: Agree. That would be an excellent next step (alerts) in a separate PR.
Probably is also wise to have the metrics up and running and see how they perform.
